2025-03-11 Alive Blogpost

Annacaro22 opened the pull request:
@katherinewu312 
@InnovativeInventor 
@calciiium

Annacaro22 committed:
2025-03-11 Blogpost

sampsyo commented on content/blog/2025-03-11-Alive-Peephole-Optimizations:
```diff
@@ -0,0 +1,62 @@
+name = "Stephanie Ma"
++++
+
+# Alive: Provably Correct Peephole Optimizations
```
You don't need this header; the blog post page already displays your title. (You can preview what this looks like by running Zola yourself.)

sampsyo commented on content/blog/2025-03-11-Alive-Peephole-Optimizations:
```diff
@@ -0,0 +1,62 @@
++++
+
+# Alive: Provably Correct Peephole Optimizations
+A blog post about the paper ["Provably Correct Peephole Optimizations With Alive"](https://dl.acm.org/doi/pdf/10.1145/2737924.2737965).
```
I don't think you need this sentence… you can just link to the paper from the first sentence of your overview.

sampsyo commented on content/blog/2025-03-11-Alive-Peephole-Optimizations:
```diff
@@ -0,0 +1,62 @@
+# Alive: Provably Correct Peephole Optimizations
+A blog post about the paper ["Provably Correct Peephole Optimizations With Alive"](https://dl.acm.org/doi/pdf/10.1145/2737924.2737965).
+
+## Background / Vocab Terms, For the Uninitiated:
+
+* [LLVM](https://llvm.org/)- An SSA-based intermediate representation used in many real world contexts, especially with C programs.
+* [DSLs](https://www.geeksforgeeks.org/domain-specific-languages-in-scala/)- Domain Specific Languages, which are small programming languages specially designed for different domains or areas of study.
+* [Peephole Optimizations](https://www.geeksforgeeks.org/peephole-optimization-in-compiler-design/)- A small, simple optimization, replacing some inefficient instruction with a more efficient one, e.g. replacing a multiply instruction with a shift left.
+* [SMT solvers](https://ocamlpro.github.io/verification_for_dummies/smt/)- Solvers which take in a mathematical formula as input, and output valid variable assignments to make the expression return true (or satisfiable), when possible. These are essentially a generalization of SAT solvers: SMT formulas include variables representing integers, strings, arrays, etc. along with constraints and opera ...
```
It's a tiny bit awkward to start with a "disembodied" list of terms before you jump into the prose. If you came across this blog post and wanted to learn about the Alive work, it might be kinda weird to have to start with a glossary. Do you think you might be able to just integrate the links you have here into the main text? For example, you can probably just link to "peephole optimizations" from the first occurrence of that phrase. And I don't think you need to define LLVM.

sampsyo commented on content/blog/2025-03-11-Alive-Peephole-Optimizations:
```diff
@@ -0,0 +1,62 @@
+## Alive:
+The main contribution of this work was the introduction of Alive, a verification framework and DSL for expressing peephole optimizations. As input, Alive takes in a DSL that expresses a certain class of peephole optimizations. As output, Alive either verifies peephole optimizations and provides LLVM code to perform the optimization, or provides a counterexample that witnesses a miscompile. Alive  ...
+
+The paper runs through the typing rules of Alive, and the features of the tool that non-C programmers may be unfamiliar with. For example, LLVM contains support for both undefined values, which represent undefined behavior and thus let the compiler pick a convenient value for each appearance of *undef*, to allow efficient optimizations, as well as poison values, which is similar to an undefined v ...
+
+These values representing undefined behavior allow the developers of Alive to codify correctness; when correctness is verified, it means that, for some given source expression, and the target expression that exists post-optimization, that the target expression is defined whenever the source is defined, the target is poison free when the source is poison free, and, if the source is defined and poi ...
```
I'd argue that even C programmers are not familiar with these concepts. They are pretty LLVM-specific ideas.

May I recommend trying to more clearly delineate why readers should care about these two paragraphs? In particular, starting with "The paper runs through the typing rules…" doesn't tell us much about what the insight is supposed to be here, or why this is the very second thing we're learning about the entire Alive system. If you believe this is indeed the second-most-urgent thing you want readers to learn, maybe start with an opening that more directly addresses what this is about: namely, that verifying equivalence might be hard enough _without_ undefinedness, but it's much harder because LLVM has undefined behavior that can both enable and prohibit optimizations. So much of the Alive paper is concerned with accurately modeling all of that subtlety.

sampsyo commented on content/blog/2025-03-11-Alive-Peephole-Optimizations:
```diff
@@ -0,0 +1,62 @@
+
+In order to do this verification procedure, SMT solvers are employed to see if each given (source, target) pair meets the correctness conditions outlined using the undefined and poison values. If an optimization fails the correctness check, then a counterexample is generated automatically by Alive as well, typically a four or eight bitwidth counterexample as those are the easiest to understand fo ...
+
+The authors display the real-world usefulness of their tool by testing it on the LLVM InstCombine transformations, in which they found multiple bugs, with the buggiest file resulting in *six of fourty-four*, or 14%, errors found by Alive. They also tested their tool on proposed LLVM patches, successfully proving correctness of LLVM patches before the PR was accepted, or locating bugs in the code  ...
```
14% of what? I think I've lost track of the denominator here (and whether I'm supposed to think that 14% is high or low).

sampsyo commented on content/blog/2025-03-11-Alive-Peephole-Optimizations:
```diff
@@ -0,0 +1,62 @@
+The authors display the real-world usefulness of their tool by testing it on the LLVM InstCombine transformations, in which they found multiple bugs, with the buggiest file resulting in *six of fourty-four*, or 14%, errors found by Alive. They also tested their tool on proposed LLVM patches, successfully proving correctness of LLVM patches before the PR was accepted, or locating bugs in the code  ...
+
+## Follow-up work regarding Alive:
+There have been several variations of Alive following this paper, including floating-point support by Alive-FP, and precondition inference for peephole optimizations expressed in Alive by Alive-Infer.
```
When you refer to other work, please use hyperlinks. You can link either to the papers for Alive-FP, Alive-Infer, Alive2, etc., or link to the open-source repositories. I imagine that any reader will want to know more, and by giving them a link, you can make it obvious where they need to go.

sampsyo commented on content/blog/2025-03-11-Alive-Peephole-Optimizations:
```diff
@@ -0,0 +1,62 @@
+## Follow-up work regarding Alive:
+There have been several variations of Alive following this paper, including floating-point support by Alive-FP, and precondition inference for peephole optimizations expressed in Alive by Alive-Infer.
+
+Six years after the release of this paper, in 2021, Alive was updated to the tool Alive 2, which expands Alive to include loop handling (by partially unrolling loops) and goes beyond simple local optimizations to handle function calls as well. It also uses translation validation which takes two IR files and checks whether the translation from one to the other is correct, so that the optimization  ...
```
I know this is a ridiculous nitpick, but I believe it's "Alive2" and not "Alive 2".

sampsyo commented on content/blog/2025-03-11-Alive-Peephole-Optimizations:
```diff
@@ -0,0 +1,62 @@
+
+We also think it makes a lot of sense to design Alive with LLVM in mind, to find an already thriving developer community and create a tool specifically for them, rather than create something radically new and different that people are then resistant to actually use in real life. Tools with a concrete audience in mind are usually better loved and more impactful overall than even the most groundbre ...
+
+However, there were a lot of concerns raised in the online discussion about both the scope/usefulness and correctness of the DSL. For example, with the scope of Alive only covering Peephole optimizations, which are small, simple cases that can be easily hand-checked, the sentiment kept popping up: how useful is the tool, really? There was a lot of interest in trying to expand a verification tool  ...
```
> Peephole optimizations, which are small, simple cases that can be easily hand-checked

I would argue that the paper provides plenty of examples where peephole optimizations _cannot_ feasibly be hand-checked, or at least cases where humans got them wrong despite attempting to hand-check them.

sampsyo replied:
Similarly, please steer clear of "useless": I don't think it's worth representing the position that verifying peephole optimizations is truly useless. I think what you mean here is that there may be other transformations that matter even more (despite the fact that InstCombine is disproportionately buggy, pound for pound).

sampsyo commented on content/blog/2025-03-11-Alive-Peephole-Optimizations:
```diff
@@ -0,0 +1,62 @@
+
+However, there were a lot of concerns raised in the online discussion about both the scope/usefulness and correctness of the DSL. For example, with the scope of Alive only covering Peephole optimizations, which are small, simple cases that can be easily hand-checked, the sentiment kept popping up: how useful is the tool, really? There was a lot of interest in trying to expand a verification tool  ...
+
+There were also some concerns about correctness in Alive’s verification procedures, since they rely on the use of SMT solvers, which have known bugs in them. However, again in in-class discussion, we couldn’t come up with any better solution; to paraphrase Prof. Sampson in our in-class discussion, “It seems to me that SMT solvers are the worst solution, besides all the other solutions.”
```
You may want to link to one of those papers about SMT solver bugs to substantiate your claim here. (It could be the same link as the one you use below.)

sampsyo commented on content/blog/2025-03-11-Alive-Peephole-Optimizations:
```diff
@@ -0,0 +1,62 @@
+Overall, while there were some concerns raised about Alive in the discussion posts, we concluded in class that these concerns were either impossible to solve or not that big of a deal; Alive is one tool in the compiler developer’s toolbox, useful for debugging with its counterexample generation, and useful in that it provides guarantee of some correctness. Even if it’s not the magic bullet to aut ...
+
+
+## Discussion Questions, taken from the Github conversations about the article:
```
An external reader probably doesn't care that these discussions happened on GitHub. Maybe just refer to, like, "the class's discussion" or something?

sampsyo commented on content/blog/2025-03-11-Alive-Peephole-Optimizations:
```diff
@@ -0,0 +1,62 @@
+
+## Discussion Questions, taken from the Github conversations about the article:
+
+* Many compiler passes are not as simple as peephole optimizations, spanning basic blocks if not function calls. Alive does not support control flow analysis, but could a similar solver-based architecture be used to verify control flow modifying passes? How?
```
FWIW, Alive2's translation validation does support control flow.

sampsyo commented on content/blog/2025-03-11-Alive-Peephole-Optimizations:
```diff
@@ -0,0 +1,62 @@
+* Many compiler passes are not as simple as peephole optimizations, spanning basic blocks if not function calls. Alive does not support control flow analysis, but could a similar solver-based architecture be used to verify control flow modifying passes? How?
+* To what extent can we trust SMT solvers to check correctness of program transformations? It seems Alive defers the proof obligation to the SMT solver, although it is well-known that SMT solvers have bugs — there’s a [line of research](https://testsmt.github.io/) showing that popular SMT solvers have >500 soundness bugs. With this in mind, is it still ok to rely on SMT solvers? If not, how they  ...
+* It's good for the developers that Alive is designed to resemble LLVM as much as possible, but does it make it hard to design or limit the expressiveness of automated verification tool like Alive in some way?
+* [Optional] What’s the best way to ensure correctness for sources with floats and aggregate types? Alive currently does not implement these. [Alive FP]
```
What does "[Optional]" mean here? And why does this line have "[Alive FP]" at the end? (Maybe this section just needs some tweaking for this context, i.e., the externally-facing blog post instead of the internally-facing class discussion.)

sampsyo reviewed (changes requested):
This is looking reasonably complete. You have a pretty good run-down of some interesting discussion outcomes. I have a few minor- to moderate-level suggestions in this review.

Annacaro22 committed:
Alive Blogpost Update

sampsyo commented:
Hi, @Annacaro22 @katherinewu312 @InnovativeInventor @calciiium! I would love to publish your blog post. Can you please wrap up the revisions discussed above so I can hit the green button?

Annacaro22 commented:
Hi! Sorry, I thought I pushed a new version of our blogpost that addressed the comments you left (the "Alive Blogpost Update" commit); was there something else I needed to do?

Thanks!

sampsyo commented:
Oh great! No, I think you're all done—I just failed to notice the new commit. Looks great!
